### PR TITLE
arch/cortex-m0: hard_fault_handler: fix incorrect return to PSP stack

### PR DESCRIPTION
**Pull request by lschuermann**
### Pull Request Overview

This pull request is the sibling of #3798 for Cortex-M0 (ARM v6m). It fixes a critical bug in the ARMv6m hard-fault handler. Prior to this fix, it is possible (depending on the compiler) that the hard-fault handler may return to an application's context (the PSP stack), but *in privileged handler mode*. This effectively and trivially allows an application to execute arbitrary code at kernel privileges. For a more in-depth description of this issue, see #3798.

### Testing Strategy

This pull request needs testing.

### TODO or Help Wanted

N/A

### Documentation Updated

- [~] ~Updated the relevant files in `/docs`, or~ no updates are required.

### Formatting

- [x] Ran `make prepush`.

### CODE REVIEW

**Review by bradjc — APPROVED** [review]

**Comment by bradjc** [comment]
It's possible we need this change for #3802 as well. I'm not sure exactly how to resolve the current build error.

**Comment by lschuermann** [comment]
> I'm not sure exactly how to resolve the current build error.

Assuming you mean this one?

```
   = note: rust-lld: error: /home/runner/work/tock/tock/target/thumbv6m-none-eabi/release/deps/raspberry_pi_pico-eef32a0b3c7dcaeb.raspberry_pi_pico.20f617c34f0755ad-cgu.0.rcgu.o:(.hard_fault_handler+0x16): relocation R_ARM_THM_JUMP11 out of range: -32429 is not in [-2048, 2047]; references 'cortexm0::hard_fault_handler_continued'
          >>> referenced by raspberry_pi_pico.20f617c34f0755ad-cgu.0
          >>> defined in /home/runner/work/tock/tock/target/thumbv6m-none-eabi/release/deps/raspberry_pi_pico-eef32a0b3c7dcaeb.raspberry_pi_pico.20f617c34f0755ad-cgu.0.rcgu.o
```

I think this is telling us that the relative offset of the `_continued` function doesn't fit into the immediate argument of the branch instruction in the hard fault handler. On the one hand, I believe this should be an easy fix (just explicitly load this symbol first, then jump to register). On the other hand, this makes me more uneasy about #3802 generally. It seems that using `#[naked]` gave the compiler and/or linker more flexibility w.r.t. making such locality optimizations that allow these relative branches between otherwise unrelated code blocks. And using global assembly somehow forces placement constraints that make these optimizations impossible. I'm worried that this is going to have more general fallout for compiler optimizations we rely on, or at least like to have.

**Comment by bradjc** [comment]
> On the other hand, this makes me more uneasy about #3802 generally. It seems that using `#[naked]` gave the compiler and/or linker more flexibility w.r.t. making such locality optimizations that allow these relative branches between otherwise unrelated code blocks. And using global assembly somehow forces placement constraints that make these optimizations impossible. I'm worried that this is going to have more general fallout for compiler optimizations we rely on, or at least like to have.

But the other could be true too? Perhaps we have unnecessarily verbose assembly elsewhere because of a compile error when using `naked` because of how the compiler placed those functions?

**Comment by lschuermann** [comment]
> But the other could be true too? Perhaps we have unnecessarily verbose assembly elsewhere because of a compile error when using `naked` because of how the compiler placed those functions?

I think I follow. I guess really my takeaway is that I'm slightly surprised that this switch of #3802 makes any tangible difference at all, and we should be looking at the binary size metrics carefully.

**Comment by lschuermann** [comment]
Let me try to debug that issue and (hopefully manage to) test this fix on a Pi Pico today.

**Comment by lschuermann** [comment]
I have managed to get Tock running on a Pi Pico board. By setting a break-point to the hard-fault-handler entry and single-stepping through these instructions, I verified that prior to this fix we jump back to the application after a hard fault, but in privileged mode. After this fix, we jump back to the `switch_to_process` handler and mark the application as faulted.

What makes this bug so tricky to spot in the first place is that the application is still marked as faulted. However, it's marked as faulted only on a subsequent system call, when `APP_HARD_FAULT` is read back.

https://github.com/tock/tock/blob/d150f8802101d345ba47fd3d88d15591d2550cc4/arch/cortex-m/src/syscall.rs#L265-L268

In privileged mode the application could simply overwrite this bit though.

tl;dr: This is good to go!

**Comment by lschuermann** [comment]
@bradjc Lets fix the issue with #3802 in that PR, given its unrelated to these changes.
